File: backend/pkg/api/storage/handlers.go
package storage

import (
	"io"
	"log"

	"backend/pkg/api/dto"
	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	coreStorage "backend/pkg/core/storage"

	"github.com/gofiber/fiber/v2"
)

// ListFiles handles GET /storage/files
// @Summary List files
// @Description Get a list of files in storage
// @Tags storage
// @Accept json
// @Produce json
// @Param prefix query string false "File prefix filter"
// @Success 200 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /storage/files [get]
func ListFiles(c *fiber.Ctx) error {
	prefix := c.Query("prefix", "")
	log.Printf("[Storage] ListFiles: prefix=%q", prefix)

	result, err := coreStorage.ListFiles(c.Context(), prefix)
	if err != nil {
		log.Printf("[Storage] ListFiles: error: %v", err)
		return response.Error(c, err)
	}
	log.Printf("[Storage] ListFiles: prefix=%q files=%d folders=%d", prefix, len(result.Files), len(result.Folders))
	return response.Success(c, result)
}

// UploadFile handles POST /storage/upload
// @Summary Upload file
// @Description Upload a file to storage
// @Tags storage
// @Accept multipart/form-data
// @Produce json
// @Param key formData string true "File key/path"
// @Param file formData file true "File to upload"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean,url=string,key=string}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /storage/upload [post]
func UploadFile(c *fiber.Ctx) error {
	contentType := c.Get("Content-Type")
	log.Printf("[Storage] UploadFile: content-type=%q method=%s", contentType, c.Method())

	// Get the key from form data
	key := c.FormValue("key")
	log.Printf("[Storage] UploadFile: key=%q", key)
	if key == "" {
		log.Printf("[Storage] UploadFile: key is empty, returning 400")
		return response.Error(c, core.InvalidInputError("File key is required"))
	}

	// Get the file from form data
	fileHeader, err := c.FormFile("file")
	if err != nil {
		log.Printf("[Storage] UploadFile: failed to get file from form: %v", err)
		return response.Error(c, core.InvalidInputError("File is required"))
	}
	log.Printf("[Storage] UploadFile: file=%q size=%d", fileHeader.Filename, fileHeader.Size)

	// Open the file
	file, err := fileHeader.Open()
	if err != nil {
		log.Printf("[Storage] UploadFile: failed to open file: %v", err)
		return response.Error(c, err)
	}
	defer file.Close()

	// Read file contents
	data, err := io.ReadAll(file)
	if err != nil {
		log.Printf("[Storage] UploadFile: failed to read file: %v", err)
		return response.Error(c, err)
	}

	// Upload to storage
	log.Printf("[Storage] UploadFile: uploading key=%q (%d bytes)", key, len(data))
	if err := coreStorage.UploadFile(c.Context(), key, data); err != nil {
		log.Printf("[Storage] UploadFile: storage upload failed: %v", err)
		return response.Error(c, err)
	}

	// Return the URL of the uploaded file
	url := coreStorage.GetURLPrefix() + "/" + key
	log.Printf("[Storage] UploadFile: success key=%q url=%q", key, url)
	return response.Success(c, fiber.Map{
		"success": true,
		"url":     url,
		"key":     key,
	})
}

// DeleteFile handles DELETE /storage/:key
// @Summary Delete file
// @Description Delete a file from storage
// @Tags storage
// @Accept json
// @Produce json
// @Param key path string true "File key/path"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /storage/{key} [delete]
func DeleteFile(c *fiber.Ctx) error {
	key := c.Params("key")
	log.Printf("[Storage] DeleteFile: key=%q", key)
	if key == "" {
		return response.Error(c, core.InvalidInputError("File key is required"))
	}

	if err := coreStorage.DeleteFile(c.Context(), key); err != nil {
		log.Printf("[Storage] DeleteFile: error key=%q: %v", key, err)
		return response.Error(c, err)
	}
	log.Printf("[Storage] DeleteFile: success key=%q", key)
	return response.Success(c, fiber.Map{"success": true})
}

// CreateFolder handles POST /storage/folders
// @Summary Create folder
// @Description Create a new folder in storage
// @Tags storage
// @Accept json
// @Produce json
// @Param request body dto.CreateFolderRequest true "Folder path"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /storage/folders [post]
func CreateFolder(c *fiber.Ctx) error {
	var req dto.CreateFolderRequest
	if err := c.BodyParser(&req); err != nil {
		log.Printf("[Storage] CreateFolder: failed to parse body: %v", err)
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		log.Printf("[Storage] CreateFolder: validation error: %v", err)
		return response.Error(c, err)
	}

	log.Printf("[Storage] CreateFolder: path=%q", req.Path)
	if err := coreStorage.CreateFolder(c.Context(), req.Path); err != nil {
		log.Printf("[Storage] CreateFolder: error path=%q: %v", req.Path, err)
		return response.Error(c, err)
	}
	log.Printf("[Storage] CreateFolder: success path=%q", req.Path)
	return response.Success(c, fiber.Map{"success": true})
}

// UpdateFolder handles PUT /storage/folders
// @Summary Update folder
// @Description Rename/move a folder in storage
// @Tags storage
// @Accept json
// @Produce json
// @Param request body dto.UpdateFolderRequest true "Old and new folder paths"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /storage/folders [put]
func UpdateFolder(c *fiber.Ctx) error {
	var req dto.UpdateFolderRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	if err := coreStorage.UpdateFolder(c.Context(), req.OldPath, req.NewPath); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

File: backend/src/api/storage/handlers.rs
use axum::{
    Json,
    extract::{Multipart, Path, Query, State, multipart::MultipartRejection},
};

use crate::{
    api::{auth::AuthenticatedAccount, request::JsonBody, response::SuccessResponse},
    error::AppError,
};

use super::{
    dto::{
        CreateFolderRequest, ListFilesQuery, ListFilesResponse, SuccessFlagResponse,
        UpdateFolderRequest, UploadFileRequest, UploadFileResponse,
    },
    state::StorageState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, AppError>;

#[utoipa::path(
    get,
    path = "/storage/files",
    params(ListFilesQuery),
    responses(
        (status = 200, body = SuccessResponse<ListFilesResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "storage",
    operation_id = "listStorageFiles"
)]
pub async fn list_files(
    _authenticated: AuthenticatedAccount,
    State(state): State<StorageState>,
    Query(query): Query<ListFilesQuery>,
) -> ApiResult<ListFilesResponse> {
    let result = state
        .service()?
        .list_files(query.prefix.as_deref().unwrap_or_default())
        .await?;
    Ok(Json(SuccessResponse::new(ListFilesResponse {
        files: result.files.into_iter().map(Into::into).collect(),
        folders: result.folders.into_iter().map(Into::into).collect(),
    })))
}

#[utoipa::path(
    post,
    path = "/storage/upload",
    request_body(content = UploadFileRequest, content_type = "multipart/form-data"),
    responses(
        (status = 200, body = SuccessResponse<UploadFileResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "storage",
    operation_id = "uploadStorageFile"
)]
pub async fn upload_file(
    _authenticated: AuthenticatedAccount,
    State(state): State<StorageState>,
    multipart: Result<Multipart, MultipartRejection>,
) -> ApiResult<UploadFileResponse> {
    let mut multipart =
        multipart.map_err(|_| AppError::InvalidInput("Invalid request body".to_owned()))?;
    let mut key = None;
    let mut file = None;
    while let Some(field) = multipart
        .next_field()
        .await
        .map_err(|_| AppError::InvalidInput("Invalid request body".to_owned()))?
    {
        match field.name() {
            Some("key") if key.is_none() => {
                key = Some(
                    field
                        .text()
                        .await
                        .map_err(|_| AppError::InvalidInput("Invalid request body".to_owned()))?,
                );
            }
            Some("file") if file.is_none() => {
                file = Some(
                    field
                        .bytes()
                        .await
                        .map_err(|_| AppError::Internal)?
                        .to_vec(),
                );
            }
            _ => {}
        }
    }
    let key = key
        .filter(|value| !value.is_empty())
        .ok_or_else(|| AppError::InvalidInput("File key is required".to_owned()))?;
    let file = file.ok_or_else(|| AppError::InvalidInput("File is required".to_owned()))?;
    let service = state.service()?;
    service.upload_file(&key, file).await?;
    Ok(Json(SuccessResponse::new(UploadFileResponse {
        success: true,
        url: format!("{}/{}", service.url_prefix(), key),
        key,
    })))
}

#[utoipa::path(
    delete,
    path = "/storage/{key}",
    params(("key" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "storage",
    operation_id = "deleteStorageFile"
)]
pub async fn delete_file(
    _authenticated: AuthenticatedAccount,
    State(state): State<StorageState>,
    Path(key): Path<String>,
) -> ApiResult<SuccessFlagResponse> {
    if key.is_empty() {
        return Err(AppError::InvalidInput("File key is required".to_owned()));
    }
    state.service()?.delete_file(&key).await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}

#[utoipa::path(
    post,
    path = "/storage/folders",
    request_body = CreateFolderRequest,
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "storage",
    operation_id = "createStorageFolder"
)]
pub async fn create_folder(
    _authenticated: AuthenticatedAccount,
    State(state): State<StorageState>,
    JsonBody(request): JsonBody<CreateFolderRequest>,
) -> ApiResult<SuccessFlagResponse> {
    if request.path.is_empty() {
        return Err(AppError::InvalidInput(
            "path is a required field".to_owned(),
        ));
    }
    state.service()?.create_folder(&request.path).await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}

#[utoipa::path(
    put,
    path = "/storage/folders",
    request_body = UpdateFolderRequest,
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "storage",
    operation_id = "updateStorageFolder"
)]
pub async fn update_folder(
    _authenticated: AuthenticatedAccount,
    State(state): State<StorageState>,
    JsonBody(request): JsonBody<UpdateFolderRequest>,
) -> ApiResult<SuccessFlagResponse> {
    if request.old_path.is_empty() {
        return Err(AppError::InvalidInput(
            "oldPath is a required field".to_owned(),
        ));
    }
    if request.new_path.is_empty() {
        return Err(AppError::InvalidInput(
            "newPath is a required field".to_owned(),
        ));
    }
    state
        .service()?
        .update_folder(&request.old_path, &request.new_path)
        .await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}
